cloud/http.go

```diff
@@ -109,9 +110,10 @@ func (c *client) doCall(ctx context.Context, method, url string, opts ...request
 	body := []byte{}
 	var callErr error
 	duration := time.Millisecond * 100
+	reqID := uuid.NewString()
```

Comment:
Shouldn't the request ID be created by the server and then returned as part of any responses (especially as part of error responses)?

Reply:
It can be done that way, but I think it makes more sense for the client to generate it.

One reason is idempotency - the client may retry failed requests using the same request-id, and the server can be coded to ensure it doesn't duplicate things it shouldn't by distinguishing between a retry vs an entirely new request. For example, [Stripe APIs work this way](https://stripe.com/docs/api/idempotent_requests), and [Google suggests doing so in gRPC](https://cloud.google.com/apis/design/design_patterns) (although I think their suggestion is in the request message, vs metadata, but close enough).

It's sometimes documented as a standard-ish HTTP header for correlating server logs to client requests, e.g. [heroku](https://http.dev/x-request-id) and [other blogs](https://http.dev/x-request-id) mention it.

Also similar to how distributed tracing works, where a tracked request usually begins at the client in the form of some unique ID, before it's passed through various backend services. P.S. Our backend also passes this request ID to subsequent internal microservices, so we can do some of that tracing via our logs.

Reply:
I'm in favour of generating it client side (see my comment regarding https://github.com/earthly/earthly/pull/2404/files#r1021874460 )

Is there any way we can make use of `withReqID()` here? Having a single code path for generating (and potentially overriding) the request ID would be cleaner.

Reply:
Sounds good then - I was a bit worried about someone being able to spam us. But then again, it's hard to guess other user's UUIDs to create any real issues.